File: apps/worker/worker/evaluations/intervals.py
"""Deterministic employment-interval arithmetic over extracted dates.

The extraction model may identify spans, but durations, overlaps, and
totals are computed here so score-relevant date math never depends on
model output.
"""

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any, cast


@dataclass(frozen=True)
class MonthInterval:
	# Half-open span of calendar months as year * 12 + month indices.
	start: int
	end: int

	@property
	def months(self) -> int:
		return max(0, self.end - self.start)


@dataclass(frozen=True)
class EmploymentIntervals:
	valid: tuple[MonthInterval, ...]
	merged: tuple[MonthInterval, ...]
	overlap_count: int

	@property
	def total_months(self) -> int:
		return sum(interval.months for interval in self.merged)


def month_index(value: object) -> int | None:
	if not isinstance(value, str):
		return None
	parts = value.split("-")
	try:
		year = int(parts[0])
		month = int(parts[1]) if len(parts) > 1 else 1
	except ValueError:
		return None
	if not 1 <= month <= 12:
		return None
	return year * 12 + month


def employment_intervals(
	entries: object, *, now: datetime | None = None
) -> EmploymentIntervals:
	"""Build valid dated intervals from extracted employment facts.

	Entries without a start, with an end before their start, or otherwise
	unparsable dates are dropped rather than guessed. A current role
	without an end date runs to the reference month so totals stay
	deterministic for a given evaluation time.
	"""
	raw_entries = cast(list[object], entries) if isinstance(entries, list) else []
	reference = now or datetime.now(UTC)
	current_month = reference.year * 12 + reference.month
	valid: list[MonthInterval] = []
	for item in raw_entries:
		if not isinstance(item, Mapping):
			continue
		entry = cast(Mapping[str, Any], item)
		start = month_index(entry.get("startDate"))
		end = month_index(entry.get("endDate"))
		if end is None and entry.get("isCurrent") is True:
			end = current_month
		if start is None or end is None or end < start:
			continue
		valid.append(MonthInterval(start, end))
	return EmploymentIntervals(
		valid=tuple(valid),
		merged=tuple(_merged(valid)),
		overlap_count=sum(
			left.start <= right.end and right.start <= left.end
			for index, left in enumerate(valid)
			for right in valid[index + 1 :]
		),
	)


def _merged(intervals: Sequence[MonthInterval]) -> list[MonthInterval]:
	ordered = sorted(intervals, key=lambda interval: (interval.start, interval.end))
	merged: list[MonthInterval] = []
	for interval in ordered:
		if merged and interval.start <= merged[-1].end:
			previous = merged[-1]
			merged[-1] = MonthInterval(previous.start, max(previous.end, interval.end))
			continue
		merged.append(interval)
	return merged

File: apps/worker-new/src/domain/intervals.ts
export interface MonthInterval {
	readonly start: number;
	readonly end: number;
}

export interface EmploymentIntervals {
	readonly valid: readonly MonthInterval[];
	readonly merged: readonly MonthInterval[];
	readonly overlapCount: number;
	readonly totalMonths: number;
}

export function monthIndex(value: unknown): number | null {
	if (typeof value !== "string") return null;
	const parts = value.split("-");
	const year = Number(parts[0]);
	const month = parts.length > 1 ? Number(parts[1]) : 1;
	if (!Number.isInteger(year) || !Number.isInteger(month) || month < 1 || month > 12) return null;
	return year * 12 + month;
}

export function employmentIntervals(entries: unknown, now = new Date()): EmploymentIntervals {
	const valid: MonthInterval[] = [];
	const currentMonth = now.getUTCFullYear() * 12 + now.getUTCMonth() + 1;
	if (Array.isArray(entries)) {
		for (const item of entries) {
			if (!isRecord(item)) continue;
			const start = monthIndex(item["startDate"]);
			let end = monthIndex(item["endDate"]);
			if (end === null && item["isCurrent"] === true) end = currentMonth;
			if (start === null || end === null || end < start) continue;
			valid.push({ start, end });
		}
	}
	const merged = mergeIntervals(valid);
	let overlapCount = 0;
	for (let index = 0; index < valid.length; index += 1) {
		const left = valid[index];
		if (!left) continue;
		for (const right of valid.slice(index + 1)) {
			if (left.start <= right.end && right.start <= left.end) overlapCount += 1;
		}
	}
	return { valid, merged, overlapCount, totalMonths: merged.reduce((sum, item) => sum + Math.max(0, item.end - item.start), 0) };
}

function mergeIntervals(intervals: readonly MonthInterval[]): MonthInterval[] {
	const ordered = [...intervals].sort((left, right) => left.start - right.start || left.end - right.end);
	const merged: MonthInterval[] = [];
	for (const interval of ordered) {
		const previous = merged.at(-1);
		if (previous && interval.start <= previous.end) {
			merged[merged.length - 1] = { start: previous.start, end: Math.max(previous.end, interval.end) };
		} else {
			merged.push(interval);
		}
	}
	return merged;
}

function isRecord(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}
